### PR TITLE
Implement National Focus Loading

### PR DESCRIPTION
**Pull request by BrickPi**
Loads `common/national_focus.txt`
Adds launch.json for MSVC debugging
Fixes bug in `expect_modifier_value_and_keys_and_default()`

Needs testing.

### CODE REVIEW

File: src/openvic-simulation/politics/NationalFocus.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+			NationalFocus::pop_promotion_map_t promotions;
+			NationalFocus::party_loyalty_map_t loyalties;
+			NationalFocus::production_map_t production;
+
+			Ideology const* last_specified_ideology = nullptr; // weird, I know
+
+			bool ret = modifier_manager.expect_modifier_value_and_keys_and_default(
+				move_variable_callback(modifiers),
+				[&promotions, &pop_manager, &production, &good_manager, &modifiers, &modifier_manager](std::string_view key, ast::NodeCPtr value) -> bool {
+					if (pop_manager.has_pop_type_identifier(key)) {
```
Comment (Hop311) [inline src/openvic-simulation/politics/NationalFocus.cpp:66]: Under the hood this searches for the pop type pointer and checks it's not null, if you get the pointer and do the check yourself you can avoid an unnecessary second search. (Same thing for the goods check beneath this)

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (Hop311) [inline .vscode/launch.json:1]: Did you mean to include this file?

Comment (BrickPi) [reply .vscode/launch.json:1]: Yeah, it's a generic VSCode debugging setup for windows.

File: src/openvic-simulation/politics/NationalFocus.hpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	struct NationalFocus : HasIdentifier {
+		friend struct NationalFocusManager;
+
+	public:
+		using pop_promotion_map_t = std::map<PopType const*, fixed_point_t>;
+		using party_loyalty_map_t = std::map<Ideology const*, fixed_point_t>;
+		using production_map_t = std::map<Good const*, fixed_point_t>;
+
+	private:
+		std::string PROPERTY(icon);
```
Comment (Hop311) [inline src/openvic-simulation/politics/NationalFocus.hpp:31]: This should be a uint*_t (width is up to you), and check in add_national_focus should be that it's non-zero (the 0-th icon is the "no/unset national focus" one)

**Review by Hop311 — COMMENTED** [review]
Needs one more small change I just noticed, also it's failing when I test it but with no error message beyond the one in Dataloader.cpp, investigating

File: src/openvic-simulation/politics/NationalFocus.cpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+						return ret;
+					}
+					Good const* good = good_manager.get_good_by_identifier(key);
+					if (good != nullptr) {
+						fixed_point_t boost;
+						bool ret = expect_fixed_point(assign_variable_callback(boost))(value);
+						production[good] = boost;
+						return ret;
+					}
+					return false;
```
Comment (Hop311) [inline src/openvic-simulation/politics/NationalFocus.cpp:89]:
```suggestion
					return key_value_invalid_callback(key, value);
```
OK this revealed the error, invalid key `railroads` (modifier affecting capitalist chance to build railroads). Add `ret &= add_modifier_effect("railroads", true);` to `Modifier.cpp` and everything loads successfully.
In general everywhere you `return false;` should have an error message so we know where the failure occurred, here `key_value_invalid_callback` is basically a prepackaged error message.

**Review by Hop311 — APPROVED** [review]
Great work Brick 👍

**Pull request merged**